pkg/skaffold/docker/client.go

```diff
@@ -65,29 +71,46 @@ type Config interface {
 // NewAPIClientImpl guesses the docker client to use based on current Kubernetes context.
 func NewAPIClientImpl(ctx context.Context, cfg Config) (LocalDaemon, error) {
 	dockerAPIClientOnce.Do(func() {
-		env, apiClient, err := newAPIClient(ctx, cfg.GetKubeContext(), cfg.MinikubeProfile())
+		env, apiClient, err := newAPIClient(ctx, cfg.GetKubeContext())
 		dockerAPIClient = NewLocalDaemon(apiClient, env, cfg.Prune(), cfg)
 		dockerAPIClientErr = err
 	})
 
 	return dockerAPIClient, dockerAPIClientErr
 }
 
-// TODO(https://github.com/GoogleContainerTools/skaffold/issues/3668):
-// remove minikubeProfile from here and instead detect it by matching the
-// kubecontext API Server to minikube profiles
-
 // newAPIClient guesses the docker client to use based on current Kubernetes context.
-func newAPIClient(ctx context.Context, kubeContext string, minikubeProfile string) ([]string, client.CommonAPIClient, error) {
-	if minikubeProfile != "" { // skip validation if explicitly specifying minikubeProfile.
-		return newMinikubeAPIClient(ctx, minikubeProfile)
-	}
-	if cluster.GetClient().IsMinikube(ctx, kubeContext) {
-		return newMinikubeAPIClient(ctx, kubeContext)
+func newAPIClient(ctx context.Context, kubecontext string) ([]string, client.CommonAPIClient, error) {
+	if cluster.GetClient().IsMinikube(ctx, kubecontext) {
+		profile, driver, err := getMinikubeProfile(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+		return newMinikubeAPIClient(ctx, profile, driver)
 	}
+
 	return newEnvAPIClient()
 }
 
+func getMinikubeProfile(ctx context.Context) (string, string, error) {
```

Comment:
Although the original implementation assumes that kubecontext name is never changed after a minikube profile is created and kubecontext(name) is the same as profile name, this is more likely to be true than the current one which assumes that  kubectl is configured to connect to the first profile. So I think we could use this method to get the driver with profile(kubecontext) from the result of ``minikube profile list -o json``,  I cannot find the command to get the detail of a profile, although we can ``cat ~/.minikube/profiles/{{profile}}/config.json`` to get the profile detail directly, that seems dangerous, I wouldn't suggest to do that.